### PR TITLE
Serialization based on reader/writer

### PR DESCRIPTION
**Pull request by VictorMikheev**

### CODE REVIEW

File: src/main/scala/scorex/util/Extensions.scala [hunk 1]
```diff
@@ -0,0 +1,167 @@
+package scorex.util
+
+import java.nio.ByteBuffer
+
+import scala.collection.generic.CanBuildFrom
+import scala.reflect.ClassTag
+
+object Extensions {
+
+  implicit class ByteOps(b: Byte) {
```
Comment (ergomorphic) [inline src/main/scala/scorex/util/Extensions.scala]: @VictorMikheev See #8

File: src/main/scala/scorex/util/serialization/Reader.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+  type CH
+
+  def getChunk(size: Int): CH
+  /**
+    * Get a byte at current position without advancing the position.
+    * @return byte at current position
+    */
+  def peekByte(): Byte
+  def getByte(): Byte
```
Comment (ergomorphic) [inline src/main/scala/scorex/util/serialization/Reader.scala:41]: @VictorMikheev please add blank line, otherwise the doc comment is  confusing

**Review by greenhat — COMMENTED** [review]
LGTM.

**Review by greenhat — COMMENTED** [review]
LGTM.

File: src/test/scala/scorex/util/serialization/VLQByteBufferReaderWriterImpSpecification.scala [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+          // test all paths
+          reader.getLong() shouldEqual v
+          reader.getULong() shouldEqual v
+        case v: Array[Byte] =>
+          val size = reader.getUShort()
+          reader.getBytes(size) shouldEqual v
+        case v: Array[Boolean] =>
+          val size = reader.getUShort()
+          reader.getBits(size) shouldEqual v
+        case ref@_ => fail(s"reader: unsupported value type: ${ref.getClass}");
```
Comment (oskin1) [inline src/test/scala/scorex/util/serialization/VLQByteBufferReaderWriterImpSpecification.scala]: Why not just use `case ref => `?

File: src/main/scala/scorex/util/serialization/VLQReader.scala [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  /**
+    * Decode a signed value previously ZigZag-encoded with [[VLQByteBufferWriter.encodeZigZagInt]]
+    *
+    * @see [[https://developers.google.com/protocol-buffers/docs/encoding#types]]
+    * @param n unsigned Int previously encoded with [[VLQByteBufferWriter.encodeZigZagInt]]
+    * @return signed Int
+    */
+  def decodeZigZagInt(n: Int): Int =
+  // source: http://github.com/google/protobuf/blob/a7252bf42df8f0841cf3a0c85fdbf1a5172adecb/java/core/src/main/java/com/google/protobuf/CodedInputStream.java#L553
+    (n >>> 1) ^ -(n & 1)
```
Comment (oskin1) [inline src/main/scala/scorex/util/serialization/VLQReader.scala]: Why separate function declaration apart from its body with comment line when no braces used `{}`? 
Here and in many places below.

File: src/main/scala/scorex/util/Extensions.scala (hunk 1 above)
Comment (VictorMikheev) [reply src/main/scala/scorex/util/Extensions.scala]: fixed

File: src/main/scala/scorex/util/serialization/Reader.scala (hunk 2 above)
Comment (VictorMikheev) [reply src/main/scala/scorex/util/serialization/Reader.scala:41]: fixed

File: src/test/scala/scorex/util/serialization/VLQByteBufferReaderWriterImpSpecification.scala (hunk 3 above)
Comment (VictorMikheev) [reply src/test/scala/scorex/util/serialization/VLQByteBufferReaderWriterImpSpecification.scala]: fixed

File: src/main/scala/scorex/util/serialization/VLQReader.scala (hunk 4 above)
Comment (VictorMikheev) [reply src/main/scala/scorex/util/serialization/VLQReader.scala]: fixed

**Review by greenhat — APPROVED** [review]
LGTM

File: build.sbt [hunk 5]
```diff
@@ -3,7 +3,7 @@ name := "scorex-util"
 
 organization := "org.scorexfoundation"
 
-version := "0.1.1"
+version := "0.1.3-SNAPSHOT"
```
Comment (catena2w) [inline build.sbt]: Let's use non-snapshot version. 0.1.3 will be fine

File: src/main/scala/scorex/util/Extensions.scala [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+    def multiplyExact(y: Short): Short = {
+      val r = x * y
+      if (r < Short.MinValue || r > Short.MaxValue)
+        throw new ArithmeticException("Short overflow")
+      r.toShort
+    }
+  }
+
+  implicit class IntOps(val x: Int) extends AnyVal {
```
Comment (catena2w) [inline src/main/scala/scorex/util/Extensions.scala:29]: Why there are no `addExact`, `subtractExact`, etc. for Int and Long?

File: src/main/scala/scorex/util/serialization/Serializer.scala [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package scorex.util.serialization
+
+trait Serializer[TFamily, T <: TFamily, R <: Reader, W <: Writer] {
```
Comment (oskin1) [inline src/main/scala/scorex/util/serialization/Serializer.scala:5]: Why do we need both `TFamily` and `T` here?

File: src/main/scala/scorex/util/serialization/Serializer.scala [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+  def serialize(obj: T, w: W): Unit
+  def parse(r: R): TFamily
```
Comment (oskin1) [inline src/main/scala/scorex/util/serialization/Serializer.scala:9]: Let's make some helper method with `Try[T]` return type like `parseTry` we have discussed

File: src/main/scala/scorex/util/Extensions.scala [hunk 9]
```diff
@@ -0,0 +1,167 @@
+package scorex.util
+
+import java.nio.ByteBuffer
+
+import scala.collection.generic.CanBuildFrom
+import scala.reflect.ClassTag
+
+object Extensions {
```
Comment (catena2w) [inline src/main/scala/scorex/util/Extensions.scala:6]: I propose to split this class to several smaller classes, e.g. Math with exact operations, Collections with filterMap, etc.
Also please write class comments

File: src/main/scala/scorex/util/serialization/VLQReader.scala [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    // for faster version see: http://github.com/google/protobuf/blob/a7252bf42df8f0841cf3a0c85fdbf1a5172adecb/java/core/src/main/java/com/google/protobuf/CodedInputStream.java#L1085
+    var result: Long = 0
+    var shift = 0
+    while (shift < 64) {
+      val b = getByte()
+      result = result | ((b & 0x7F).toLong << shift)
+      if ((b & 0x80) == 0) return result
+      shift += 7
+    }
+    sys.error(s"Cannot deserialize Long value. Unexpected reader $this with bytes remaining ${remaining}")
```
Comment (catena2w) [inline src/main/scala/scorex/util/serialization/VLQReader.scala:73]: We have a logger for this purposes

File: src/main/scala/scorex/util/serialization/VLQWriter.scala [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    * Encode a ZigZag-encoded 64-bit value.  ZigZag encodes signed integers
+    * into values that can be efficiently encoded with varint.  (Otherwise,
+    * negative values must be sign-extended to 64 bits to be varint encoded,
+    * thus always taking 10 bytes on the wire.)
+    *
+    * @see [[https://developers.google.com/protocol-buffers/docs/encoding#types]]
+    * @param n signed Long
+    * @return unsigned Long stored in a signed Long
+    */
+  def encodeZigZagLong(n: Long): Long =
```
Comment (catena2w) [inline src/main/scala/scorex/util/serialization/VLQWriter.scala]: @oskin1 already commented in a different place to add {} braces.
Also why the comment here is separated from method comments?

**Review by catena2w — CHANGES_REQUESTED** [review]

File: src/main/scala/scorex/util/serialization/Serializer.scala (hunk 8 above)
Comment (VictorMikheev) [reply src/main/scala/scorex/util/serialization/Serializer.scala:9]: fixed

File: src/main/scala/scorex/util/Extensions.scala (hunk 9 above)
Comment (VictorMikheev) [reply src/main/scala/scorex/util/Extensions.scala:6]: removed unused methods

File: src/main/scala/scorex/util/serialization/VLQReader.scala (hunk 10 above)
Comment (VictorMikheev) [reply src/main/scala/scorex/util/serialization/VLQReader.scala:73]: This is not for logging

File: src/main/scala/scorex/util/serialization/VLQWriter.scala (hunk 11 above)
Comment (VictorMikheev) [reply src/main/scala/scorex/util/serialization/VLQWriter.scala]: fixed

File: src/main/scala/scorex/util/serialization/Serializer.scala (hunk 7 above)
Comment (VictorMikheev) [reply src/main/scala/scorex/util/serialization/Serializer.scala:5]: used at https://github.com/Scorexfoundation/sigmastate-interpreter/blob/master/src/main/scala/sigmastate/serialization/ValueSerializer.scala

**Review by catena2w — APPROVED** [review]

**Pull request merged**